reposerver/repository/repository.go

```diff
@@ -2659,3 +2661,103 @@ func (s *Service) GetGitDirectories(_ context.Context, request *apiclient.GitDir
+		return s.checkoutRevision(gitClient, revision, false)
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to checkout git repo %s with revision %s: %v", repo.Repo, revision, err)
+	}
+	defer io.Close(closer)
+
+	files, err := gitClient.ChangedFiles(syncedRevision, revision)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to get changed files for repo %s with revision %s: %v", repo.Repo, revision, err)
+	}
+
+	changed := apppathutil.AppFilesHaveChanged(refreshPaths, files)
+
+	if !changed {
+		log.WithFields(log.Fields{"application": request.AppName, "appNamespace": request.Namespace}).Debugf("no changes found for application %s in repo %s from revision %s to revision %s", request.AppName, repo.Repo, syncedRevision, revision)
+
+		err := s.updateCachedRevision(syncedRevision, revision, request, gitClientOpts)
+		if err != nil {
+			// Only warn with the error, no need to block anything if there is a caching error.
+			log.WithFields(log.Fields{"application": request.AppName, "appNamespace": request.Namespace}).Warnf("error updating cached revision for repo %s with revision %s: %v", repo.Repo, revision, err)
+			return &apiclient.UpdateRevisionForPathsResponse{}, nil
+		}
+
+		return &apiclient.UpdateRevisionForPathsResponse{}, nil
+	}
+
+	log.WithFields(log.Fields{"application": request.AppName, "appNamespace": request.Namespace}).Debugf("changes found for application %s in repo %s from revision %s to revision %s", request.AppName, repo.Repo, syncedRevision, revision)
+	return &apiclient.UpdateRevisionForPathsResponse{Changes: true}, nil
+}
+
+func (s *Service) updateCachedRevision(oldRev, newRev string, request *apiclient.UpdateRevisionForPathsRequest, gitClientOpts git.ClientOpts) error {
+	repoRefs := make(map[string]string)
+	if request.HasMultipleSources && request.ApplicationSource.Helm != nil {
+		var err error
+		repoRefs, err = resolveReferencedSources(true, request.ApplicationSource.Helm, request.RefSources, s.newClientResolveRevision, gitClientOpts)
+		if err != nil {
+			return fmt.Errorf("failed to get repo refs for application %s in repo %s from revision %s: %w", request.AppName, request.GetRepo().Repo, request.Revision, err)
+		}
+	}
+
+	// Update revision in refSource
+	if request.HasMultipleSources && request.ApplicationSource.Helm != nil {
+		for normalizedURL := range repoRefs {
+			repoRefs[normalizedURL] = newRev
+		}
+	}
+
+	err := s.cache.SetNewRevisionManifests(newRev, oldRev, request.ApplicationSource, request.RefSources, request, request.Namespace, request.TrackingMethod, request.AppLabelKey, request.AppName, repoRefs)
+	if err != nil {
+		if err == cache.ErrCacheMiss {
+			log.WithFields(log.Fields{"application": request.AppName, "appNamespace": request.Namespace}).Debugf("manifest cache miss during comparison for application %s in repo %s from revision %s", request.AppName, request.GetRepo().Repo, oldRev)
```

Comment:
I was running some test locally and I got the following error.

The test was
- create an app with the annotation and without auto-sync
- Sync it to HEAD (08484ff536ba26aecf0103476e223297459c2fd3)
- Commit to HEAD outside the path (4bf66a93e785b7af4e56d33ec13cbf344072a8b8)
- Manual Refresh (15:02:57)
- Manual Sync (15:07:04)

```
argocd-repo-server time="2024-02-28T15:02:57Z" level=debug msg="manifest cache updated for application app2 in repo https://github.com/agaudreault/gitops-tests from revision 08484ff536ba26aecf0103476e223297459c2fd3 to revision 4bf66a93e785b7af4e56d33ec13cbf344072a8b8" appNamespace=dev application=app2
argocd-repo-server time="2024-02-28T15:07:04Z" level=debug msg="no changes found for application app2 in repo https://github.com/agaudreault/gitops-tests from revision 08484ff536ba26aecf0103476e223297459c2fd3 to revision 4bf66a93e785b7af4e56d33ec13cbf344072a8b8" appNamespace=dev application=app2
argocd-repo-server time="2024-02-28T15:07:04Z" level=warning msg="error updating cached revision for repo https://github.com/agaudreault/gitops-tests with revision 4bf66a93e785b7af4e56d33ec13cbf344072a8b8: manifest cache move error for app2: ERR no such key" appNamespace=dev application=app2
```

I wonder if we should handle the warning in a different way. From what I understand, it will warn for cache miss as long as the app differs from the HEAD, which will be a lot. 

The overall behaviour works as expected and the manifests are not regenerated during the manual refresh or sync.
The application `status.operationState.sync.revision` will be `4bf66a`, but the value of `status.operationState.syncResult.revision` and `status.sync.revision` will be `08484f` (what is displayed in the UI)

Reply:
Hmm you're right, that was previously caught by `err == cache.ErrCacheMiss` but digging into the code it looks like, unlike the Get, the new Rename for the Redis cache does not handle the cache miss and instead returns the raw Redis error. It seems to be an inconsistency because the other caches (like the in-memory) do return ErrCacheMiss on a miss, I'll see if I can update the Redis cache to handle the miss properly

Reply:
Added the cache miss in `cache/redis.go`